src/wasm-interpreter.h

```diff
@@ -263,38 +522,79 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
         continue;
       }
       auto& list = curr->list;
-      for (size_t i = 0; i < list.size(); i++) {
+      for (size_t i = blockIndex; i < list.size(); i++) {
         if (curr != top && i == 0) {
           // one of the block recursions we already handled
           continue;
         }
         flow = visit(list[i]);
+        if (flow.suspendTag) {
+          suspend(i);
+          return flow;
+        }
         if (flow.breaking()) {
           flow.clearIf(curr->name);
           break;
         }
       }
+      // If there was a value here, we only need it for the top iteration.
+      blockIndex = 0;
     }
     return flow;
   }
   Flow visitIf(If* curr) {
-    Flow flow = visit(curr->condition);
-    if (flow.breaking()) {
-      return flow;
+    // Suspend/resume support.
+    auto suspend = [&](Index resumeIndex) {
+      // To return to the same place when we resume, we stash an index:
+      //   0 - suspended in the condition
+      //   1 - suspended in the ifTrue arm
+      //   2 - suspended in the ifFalse arm
+      pushResumeEntry({Literal(int32_t(resumeIndex))}, "if");
+    };
+    Index resumeIndex = -1;
+    if (resuming) {
+      auto entry = popResumeEntry("if");
+      assert(entry.size() == 1);
+      resumeIndex = entry[0].geti32();
     }
-    if (flow.getSingleValue().geti32()) {
-      Flow flow = visit(curr->ifTrue);
-      if (!flow.breaking() && !curr->ifFalse) {
-        flow = Flow(); // if_else returns a value, but if does not
+
+    Flow flow;
+    Index condition;
+
+    if (resuming && resumeIndex > 0) {
+      // We are resuming into one of the arms. Just set the right condition.
+      condition = (resumeIndex == 1);
```

Comment:
It would be useful to have a comment above explaining that `condition` is the evaluated value of the branch condition. It took me a bit to figure out what was going on here.